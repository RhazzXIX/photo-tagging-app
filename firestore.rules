rules_version = '2';
service cloud.firestore {
  match /databases/{database}/documents {
    match /{document=**} {
      allow read;
    }
    match /leaderBoard/{gameVer} {
      allow create, update: if request.auth != null
    }
  }
}