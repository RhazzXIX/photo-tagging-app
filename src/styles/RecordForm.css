section.form {
  position: absolute;
  width: 100%;
  height: 100%;
  z-index: 1;
  display: flex;
  flex-direction: column;
  align-items: center;
}

form#record {
  position: sticky;
  top: 240px;
  background-color: aliceblue;
  padding: 20px 40px;
  border-radius: 8px;
  box-shadow: 2px 2px 10px grey;
  display: flex;
  flex-direction: column;
  gap: 10px;
}

form#record label {
  display: flex;
  flex-direction: column;
  gap: 4px;
}

form#record label input {
  padding: 4px 2px;
  border-radius: 8px;
  font-size: 1.2rem;
}

form#record button {
  width: max-content;
  padding: 4px 8px;
  border-radius: 8px;
  align-self: center;
}
