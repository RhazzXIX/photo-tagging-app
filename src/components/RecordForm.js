import parseToTimer from "../assists/parseToTimer";
import { setDoc, doc, updateDoc, getDoc } from "firebase/firestore";
import { database } from "../assists/fbConfig";
import { useState } from "react";

export default function RecordForm(props) {
  const { time, gameVer } = props;
  const [userName, setUserName] = useState("");

  async function sendDocs(e) {
    e.preventDefault();
    let totalRecord;
    const docRef = doc(database, "leaderBoard", gameVer);
    const leaderBoard = await getDoc(docRef).then((response) =>
      response.data()
    );

    if (!leaderBoard) {
      totalRecord = 0;
    } else {
      totalRecord = Object.keys(leaderBoard).length;
    }

    const playerRecord = {
      [`record${totalRecord}`]: { userName, time },
    };

    if (!leaderBoard) {
      await setDoc(docRef, playerRecord);
    } else {
      await updateDoc(docRef, playerRecord);
    }
  }

  const handleInputName = (e) => {
    setUserName(e.target.value);
  };

  const { minutes, seconds, milliseconds } = parseToTimer(time);
  return (
    <form id="record" onSubmit={sendDocs}>
      <div className="timer">
        <p>Your time is:</p>
        <h4>
          {minutes}m {seconds}s {milliseconds}ms
        </h4>
      </div>
      <label htmlFor="name">
        Name:
        <input
          type="text"
          id="name"
          value={userName}
          onChange={handleInputName}
          required
        />
      </label>
      <button type="submit" className="submit">
        Submit
      </button>
    </form>
  );
}
