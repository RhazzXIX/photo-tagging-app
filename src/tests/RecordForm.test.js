import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import RecordForm from "../components/RecordForm";
import { setDoc, doc, getDoc } from "firebase/firestore";

jest.mock("firebase/firestore", () => {
  const originalModule = jest.requireActual("firebase/firestore");

  return {
    __esModile: true,
    ...originalModule,
    doc: jest.fn(),
    setDoc: jest.fn((obj) => true),
    getDoc: jest.fn(),
  };
});

const user = userEvent.setup();
const docRef = "testRef";

describe("RecordForm component", () => {
  it("Renders on screen", () => {
    const { container } = render(<RecordForm />);

    expect(container).toMatchSnapshot();
  });
  it("Shows the user the time it took playing the game", () => {
    const { rerender } = render(<RecordForm time={5500} />);

    expect(screen.getByRole("heading").textContent).toBe("0m 5s 500ms");

    rerender(<RecordForm time={60456} />);

    expect(screen.getByRole("heading").textContent).toBe("1m 0s 456ms");
  });
  it("Lets the user input his name in the name input", async () => {
    render(<RecordForm time={5500} />);

    const nameInput = screen.getByLabelText("Name:");

    await user.type(nameInput, "Juan");

    expect(nameInput.value).toBe("Juan");
  });

  it("Lets users submit their record", async () => {
    doc.mockReturnValue(docRef);
    getDoc.mockResolvedValue({
      data() {},
    });
    render(<RecordForm time={5500} gameVer={"animeX"} />);

    const nameInput = screen.getByLabelText("Name:");
    const submitBtn = screen.getByRole("button", { name: /submit/i });

    await user.type(nameInput, "Juan");

    await user.click(submitBtn);

    expect(doc).toHaveBeenCalled();
    expect(doc.mock.calls[0][1]).toBe("leaderBoard");
    expect(doc.mock.calls[0][2]).toBe("animeX");

    expect(getDoc.mock.calls[0][0]).toBe("testRef");

    expect(setDoc).toHaveBeenCalled();
    expect(setDoc.mock.calls[0][0]).toBe("testRef");
    expect(setDoc.mock.calls[0][1]).toEqual({
      record0: {
        userName: "Juan",
        time: 5500,
      },
    });
  });
});
