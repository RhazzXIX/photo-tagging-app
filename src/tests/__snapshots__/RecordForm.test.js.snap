// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecordForm component Renders on screen 1`] = `
<div>
  <form
    id="record"
  >
    <div
      class="timer"
    >
      <p>
        Your time is:
      </p>
      <h4>
        0
        m 
        0
        s 
        0
        ms
      </h4>
    </div>
    <label
      for="name"
    >
      Name:
      <input
        id="name"
        required=""
        type="text"
        value=""
      />
    </label>
    <button
      class="submit"
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;
